File: mediapipe/tasks/cc/retrieval/universal_embedder/universal_embedder.cc
/* Copyright 2026 The MediaPipe Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "mediapipe/tasks/cc/retrieval/universal_embedder/universal_embedder.h"

#include <cstddef>
#include <cstdint>
#include <cstring>
#include <memory>
#include <optional>
#include <string>
#include <type_traits>
#include <utility>
#include <variant>
#include <vector>

#include "absl/cleanup/cleanup.h"
#include "absl/memory/memory.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "mediapipe/framework/formats/image.h"
#include "mediapipe/framework/port/status_macros.h"
#include "mediapipe/framework/timestamp.h"
#include "mediapipe/tasks/cc/components/containers/embedding_result.h"
#include "mediapipe/tasks/cc/components/utils/cosine_similarity.h"
#include "mediapipe/tasks/cc/core/base_options.h"
#include "mediapipe/tasks/cc/core/embedding_provider.h"
#include "mediapipe/tasks/cc/core/logging/factory/logging_factory.h"
#include "mediapipe/tasks/cc/core/logging/tasks_logger.h"
#include "mediapipe/tasks/cc/core/running_mode.h"
#include "runtime/core/embedding_engine_impl.h"        // from @litert_lm
#include "runtime/engine/embedding_engine.h"           // from @litert_lm
#include "runtime/engine/embedding_engine_settings.h"  // from @litert_lm
#include "runtime/engine/io_types.h"                   // from @litert_lm
#include "runtime/executor/executor_settings_base.h"   // from @litert_lm
#include "runtime/util/memory_mapped_file.h"           // from @litert_lm

namespace mediapipe::tasks::retrieval::universal_embedder {
namespace {

using ::litert::lm::EmbeddingEngineImpl;
using ::litert::lm::EmbeddingEngineSettings;
using ::litert::lm::EmbeddingOptions;
using ::litert::lm::InputAudio;
using ::litert::lm::InputData;
using ::litert::lm::InputImage;
using ::litert::lm::InputText;
using ::litert::lm::MemoryMappedFile;
using ::litert::lm::ModelAssets;

absl::StatusOr<std::string> EncodeToTga(const mediapipe::Image& image) {
  auto image_frame = image.GetImageFrameSharedPtr();
  if (image_frame == nullptr) {
    return absl::InvalidArgumentError("Image cannot be converted to CPU.");
  }
  int width = image_frame->Width();
  int height = image_frame->Height();
  int channels = image_frame->NumberOfChannels();

  if (channels != 1 && channels != 3 && channels != 4) {
    return absl::InvalidArgumentError(
        absl::StrCat("Unsupported number of image channels: ", channels));
  }

  // Initialize TGA Header
  std::uint8_t header[18] = {0};
  if (channels == 1) {
    header[2] = 3;  // Uncompressed Grayscale Image
  } else {
    header[2] = 2;  // Uncompressed True-Color Image
  }
  header[12] = width & 0xFF;  // Width LSB
  header[13] = (width >> 8) & 0xFF;
  header[14] = height & 0xFF;  // Height LSB
  header[15] = (height >> 8) & 0xFF;
  header[16] = channels * 8;  // Bits per pixel (8, 24, or 32)
  header[17] = (channels == 4)
                   ? 0x28
                   : 0x20;  // Top-Left origin flag (0x20) + 8-bit alpha (0x08)

  const std::size_t pixel_size = width * height * channels;
  std::string tga_bytes;
  tga_bytes.reserve(18 + pixel_size);
  tga_bytes.append(reinterpret_cast<const char*>(header), 18);
  tga_bytes.resize(18 + pixel_size);
  std::uint8_t* dst_pixels = reinterpret_cast<std::uint8_t*>(&tga_bytes[18]);

  const std::uint8_t* pixel_data = image_frame->PixelData();
  int width_step = image_frame->WidthStep();
  int row_size = width * channels;

  // Standard TGA specifications expect BGR or BGRA ordering (Blue, Green, Red).
  // If we have 3 or 4 channels, we need to swap Red and Blue components.
  // Grayscale has only 1 channel and does not need swapping.
  if (channels == 3) {
    for (int y = 0; y < height; ++y) {
      const std::uint8_t* src_row = pixel_data + y * width_step;
      std::uint8_t* dst_row = dst_pixels + y * row_size;
      for (int x = 0; x < width; ++x) {
        dst_row[x * 3 + 0] = src_row[x * 3 + 2];  // Blue
        dst_row[x * 3 + 1] = src_row[x * 3 + 1];  // Green
        dst_row[x * 3 + 2] = src_row[x * 3 + 0];  // Red
      }
    }
  } else if (channels == 4) {
    for (int y = 0; y < height; ++y) {
      const std::uint8_t* src_row = pixel_data + y * width_step;
      std::uint8_t* dst_row = dst_pixels + y * row_size;
      for (int x = 0; x < width; ++x) {
        dst_row[x * 4 + 0] = src_row[x * 4 + 2];  // Blue
        dst_row[x * 4 + 1] = src_row[x * 4 + 1];  // Green
        dst_row[x * 4 + 2] = src_row[x * 4 + 0];  // Red
        dst_row[x * 4 + 3] = src_row[x * 4 + 3];  // Alpha
      }
    }
  } else {
    // 1-channel Grayscale: copy directly
    for (int y = 0; y < height; ++y) {
      const std::uint8_t* src_row = pixel_data + y * width_step;
      std::uint8_t* dst_row = dst_pixels + y * row_size;
      std::memcpy(dst_row, src_row, row_size);
    }
  }

  return tga_bytes;
}

}  // namespace

UniversalEmbedder::UniversalEmbedder(
    std::shared_ptr<MemoryMappedFile> shared_mmap,
    std::unique_ptr<::litert::lm::EmbeddingEngine> engine, bool l2_normalize,
    std::unique_ptr<tasks::core::logging::TasksLogger> tasks_logger)
    : shared_mmap_(std::move(shared_mmap)),
      engine_(std::move(engine)),
      l2_normalize_(l2_normalize),
      tasks_logger_(std::move(tasks_logger)) {}

absl::StatusOr<std::unique_ptr<UniversalEmbedder>> UniversalEmbedder::Create(
    std::unique_ptr<UniversalEmbedderOptions> options) {
  auto tasks_logger = tasks::core::logging::CreateTasksLogger(
      {.task_name = "UniversalEmbedder",
       .task_running_mode = tasks::core::RunningMode::kUnspecified,
       .host_environment = options->base_options.host_environment,
       .host_system = options->base_options.host_system,
       .host_version = options->base_options.host_version,
       .app_id = options->base_options.app_id,
       .app_version = options->base_options.app_version,
       .ca_bundle_path = options->base_options.ca_bundle_path});
  ::litert::lm::Backend backend = ::litert::lm::Backend::CPU;
  if (options->base_options.delegate == tasks::core::BaseOptions::GPU) {
    backend = ::litert::lm::Backend::GPU;
  } else if (options->base_options.delegate == tasks::core::BaseOptions::NPU) {
    backend = ::litert::lm::Backend::NPU;
  }

  if (options->base_options.model_asset_path.empty()) {
    return absl::FailedPreconditionError(
        "No model asset path specified in BaseOptions.");
  }

  ABSL_ASSIGN_OR_RETURN(
      auto mmap_file,
      MemoryMappedFile::Create(options->base_options.model_asset_path));
  auto shared_mmap = std::shared_ptr<MemoryMappedFile>(std::move(mmap_file));

  ABSL_ASSIGN_OR_RETURN(
      auto model_assets,
      ModelAssets::Create(shared_mmap, options->base_options.model_asset_path));
  ABSL_ASSIGN_OR_RETURN(
      auto settings, EmbeddingEngineSettings::CreateDefault(
                         std::move(model_assets), backend, backend, backend));

  if (options->max_input_length.has_value()) {
    settings.SetMaxInputLength(options->max_input_length);
  }
  if (options->vision_tokens_per_image.has_value()) {
    settings.SetVisionTokensPerImage(options->vision_tokens_per_image);
  }

  if (options->activation_data_type.has_value()) {
    ::litert::lm::ActivationDataType litert_activation_type =
        ::litert::lm::ActivationDataType::FLOAT32;
    switch (*options->activation_data_type) {
      case ActivationDataType::FLOAT32:
        litert_activation_type = ::litert::lm::ActivationDataType::FLOAT32;
        break;
      case ActivationDataType::FLOAT16:
        litert_activation_type = ::litert::lm::ActivationDataType::FLOAT16;
        break;
      case ActivationDataType::INT16:
        litert_activation_type = ::litert::lm::ActivationDataType::INT16;
        break;
      case ActivationDataType::INT8:
        litert_activation_type = ::litert::lm::ActivationDataType::INT8;
        break;
    }
    settings.GetMutableMainExecutorSettings().SetActivationDataType(
        litert_activation_type);
    if (settings.GetMutableVisionExecutorSettings().has_value()) {
      settings.GetMutableVisionExecutorSettings()->SetActivationDataType(
          litert_activation_type);
    }
    if (settings.GetMutableAudioExecutorSettings().has_value()) {
      settings.GetMutableAudioExecutorSettings()->SetActivationDataType(
          litert_activation_type);
    }
  }

  if (options->cache_dir.has_value() && !options->cache_dir->empty()) {
    settings.GetMutableMainExecutorSettings().SetCacheDir(*options->cache_dir);
    if (settings.GetMutableVisionExecutorSettings().has_value()) {
      settings.GetMutableVisionExecutorSettings()->SetCacheDir(
          *options->cache_dir);
    }
    if (settings.GetMutableAudioExecutorSettings().has_value()) {
      settings.GetMutableAudioExecutorSettings()->SetCacheDir(
          *options->cache_dir);
    }
  }

#ifdef __EMSCRIPTEN__
  // In WebAssembly / Emscripten, virtual filesystem does not support XNNPack
  // mmap weight caching.
  settings.GetMutableMainExecutorSettings().SetDisableWeightCache(true);
  settings.GetMutableMainExecutorSettings().SetCacheDir(":nocache");
  if (settings.GetMutableVisionExecutorSettings().has_value()) {
    settings.GetMutableVisionExecutorSettings()->SetDisableWeightCache(true);
    settings.GetMutableVisionExecutorSettings()->SetCacheDir(":nocache");
  }
  if (settings.GetMutableAudioExecutorSettings().has_value()) {
    settings.GetMutableAudioExecutorSettings()->SetDisableWeightCache(true);
    settings.GetMutableAudioExecutorSettings()->SetCacheDir(":nocache");
  }
#endif

  ABSL_ASSIGN_OR_RETURN(auto engine,
                        EmbeddingEngineImpl::Create(std::move(settings)));

  tasks_logger->LogSessionStart();

  return absl::WrapUnique(
      new UniversalEmbedder(std::move(shared_mmap), std::move(engine),
                            options->l2_normalize, std::move(tasks_logger)));
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::ExecuteInference(const std::vector<InputData>& contents,
                                    absl::string_view error_message) {
  mediapipe::Timestamp task_logger_ts(tasks_logger_timestamp_++);
  tasks_logger_->RecordCpuInputArrival(task_logger_ts);
  absl::Cleanup log_invocation_end = [this, task_logger_ts] {
    tasks_logger_->RecordInvocationEnd(task_logger_ts);
  };

  EmbeddingOptions options;
  options.normalize = l2_normalize_;
  options.insert_special_tokens = true;

  ABSL_ASSIGN_OR_RETURN(auto response,
                        engine_->ComputeEmbedding(contents, options),
                        _ << error_message);

  components::containers::EmbeddingResult result;
  components::containers::Embedding embedding;
  embedding.float_embedding = std::move(response.embedding);
  embedding.head_index = 0;
  result.embeddings.push_back(std::move(embedding));

  return result;
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::EmbedText(absl::string_view text) {
  std::vector<InputData> contents;
  contents.push_back(InputText(std::string(text)));
  return ExecuteInference(contents, "Failed to compute text embedding.");
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::EmbedImage(absl::string_view image_bytes) {
  std::vector<InputData> contents;
  contents.push_back(InputImage(std::string(image_bytes)));
  return ExecuteInference(contents, "Failed to compute image embedding.");
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::EmbedImage(const mediapipe::Image& image) {
  ABSL_ASSIGN_OR_RETURN(std::string tga_bytes, EncodeToTga(image));
  return EmbedImage(tga_bytes);
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::EmbedAudio(const std::vector<float>& audio_data) {
  std::vector<InputData> contents;
  contents.push_back(InputAudio(audio_data));
  return ExecuteInference(contents, "Failed to compute audio embedding.");
}

absl::StatusOr<components::containers::EmbeddingResult>
UniversalEmbedder::EmbedContent(const std::vector<Part>& content) {
  std::vector<InputData> contents;
  for (const auto& part : content) {
    auto status = std::visit(
        [&contents](const auto& arg) -> absl::Status {
          using T = std::decay_t<decltype(arg)>;
          if constexpr (std::is_same_v<T, TextPart>) {
            contents.push_back(InputText(arg.text));
          } else if constexpr (std::is_same_v<T, ImagePart>) {
            contents.push_back(InputImage(arg.image_bytes));
          } else if constexpr (std::is_same_v<T, AudioPart>) {
            contents.push_back(InputAudio(arg.audio_data));
          }
          return absl::OkStatus();
        },
        part);
    if (!status.ok()) {
      return status;
    }
  }

  return ExecuteInference(contents, "Failed to compute content embedding.");
}

absl::StatusOr<double> UniversalEmbedder::CosineSimilarity(
    const components::containers::Embedding& u,
    const components::containers::Embedding& v) {
  return components::utils::CosineSimilarity(u, v);
}

std::unique_ptr<core::EmbeddingProvider> UniversalEmbedder::GetProvider() {
  class UniversalEmbedderProvider : public core::EmbeddingProvider {
   public:
    explicit UniversalEmbedderProvider(UniversalEmbedder* embedder)
        : embedder_(embedder) {}
    absl::StatusOr<std::optional<std::vector<float>>> EmbedContent(
        const std::vector<mediapipe::tasks::core::TaskPart>& content) override {
      std::vector<UniversalEmbedder::Part> cpp_parts;
      cpp_parts.reserve(content.size());
      for (const auto& part : content) {
        if (std::holds_alternative<mediapipe::tasks::core::TextPart>(part)) {
          cpp_parts.push_back(UniversalEmbedder::TextPart{
              .text = std::get<mediapipe::tasks::core::TextPart>(part).text});
        } else if (std::holds_alternative<mediapipe::tasks::core::ImagePart>(
                       part)) {
          cpp_parts.push_back(UniversalEmbedder::ImagePart{
              .image_bytes = std::get<mediapipe::tasks::core::ImagePart>(part)
                                 .image_bytes});
        } else if (std::holds_alternative<mediapipe::tasks::core::AudioPart>(
                       part)) {
          cpp_parts.push_back(UniversalEmbedder::AudioPart{
              .audio_data = std::get<mediapipe::tasks::core::AudioPart>(part)
                                .audio_data});
        }
      }
      auto result_status = embedder_->EmbedContent(cpp_parts);
      if (!result_status.ok()) {
        return result_status.status();
      }
      if (result_status->embeddings.empty()) {
        return std::nullopt;
      }
      return result_status->embeddings[0].float_embedding;
    }

   private:
    UniversalEmbedder* embedder_;
  };
  return std::make_unique<UniversalEmbedderProvider>(this);
}

UniversalEmbedder::~UniversalEmbedder() { Close().IgnoreError(); }

absl::Status UniversalEmbedder::Close() {
  if (tasks_logger_) {
    tasks_logger_->LogSessionEnd();
    tasks_logger_.reset();
  }
  return absl::OkStatus();
}

}  // namespace mediapipe::tasks::retrieval::universal_embedder

File: mediapipe/tasks/java/com/google/mediapipe/tasks/retrieval/universalembedder/UniversalEmbedder.java
/* Copyright 2026 The MediaPipe Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package com.google.mediapipe.tasks.retrieval.universalembedder;

import static java.nio.charset.StandardCharsets.UTF_8;

import android.content.Context;
import android.graphics.Bitmap;
import com.google.ai.edge.litertlm.Backend;
import com.google.ai.edge.litertlm.EmbeddingEngine;
import com.google.ai.edge.litertlm.EmbeddingEngineConfig;
import com.google.ai.edge.litertlm.EmbeddingOptions;
import com.google.ai.edge.litertlm.EmbeddingResponse;
import com.google.ai.edge.litertlm.InputData;
import com.google.common.collect.ImmutableList;
import com.google.mediapipe.calculator.proto.InferenceCalculatorProto.InferenceCalculatorOptions.Delegate.LiteRt;
import com.google.mediapipe.framework.MediaPipeException;
import com.google.mediapipe.framework.image.BitmapExtractor;
import com.google.mediapipe.framework.image.MPImage;
import com.google.mediapipe.tasks.components.containers.AudioData;
import com.google.mediapipe.tasks.components.containers.Embedding;
import com.google.mediapipe.tasks.components.containers.EmbeddingResult;
import com.google.mediapipe.tasks.components.utils.CosineSimilarity;
import com.google.mediapipe.tasks.core.BaseOptionsUtils;
import com.google.mediapipe.tasks.core.Delegate;
import com.google.mediapipe.tasks.core.EmbeddingProvider;
import com.google.mediapipe.tasks.core.logging.TasksStatsLogger;
import com.google.mediapipe.tasks.core.logging.TasksStatsLoggerFactory;
import com.google.mediapipe.tasks.core.proto.BaseOptionsProto;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;

/**
 * UniversalEmbedder combines the features of the existing Audio, Text, and Image embedders, while
 * adding support for video embeddings. It wraps the native LiteRT-LM EmbeddingEngine directly,
 * providing real on-device multimodal embeddings.
 */
@SuppressWarnings("EnumOrdinal")
public final class UniversalEmbedder implements AutoCloseable {

  private final UniversalEmbedderOptions options;
  private final EmbeddingEngine engine;
  private final TasksStatsLogger statsLogger;
  private final AtomicLong syntheticTimestamp = new AtomicLong(0);

  private UniversalEmbedder(Context context, UniversalEmbedderOptions options) {
    this.options = options;
    this.statsLogger =
        TasksStatsLoggerFactory.create(context, "UniversalEmbedder", /* taskRunningModeStr= */ "");
    this.statsLogger.logSessionStart();

    BaseOptionsProto.BaseOptions proto =
        BaseOptionsUtils.convertBaseOptionsToProto(options.baseOptions());

    String finalPath = null;
    Integer fd = null;
    if (proto.getModelAsset().hasFileDescriptorMeta()
        && proto.getModelAsset().getFileDescriptorMeta().getFd() > 0) {
      fd = proto.getModelAsset().getFileDescriptorMeta().getFd();
    } else if (!proto.getModelAsset().getFileName().isEmpty()) {
      finalPath = proto.getModelAsset().getFileName();
    } else {
      throw new MediaPipeException(
          MediaPipeException.StatusCode.INVALID_ARGUMENT.ordinal(),
          "Model path or file descriptor must be specified in BaseOptions.");
    }

    Backend defaultBackend = getBackend(context, proto);
    Backend backend =
        options.textDelegate().isPresent()
            ? getBackendForDelegate(context, options.textDelegate().get(), defaultBackend)
            : defaultBackend;
    Backend visionBackend =
        options.visionDelegate().isPresent()
            ? getBackendForDelegate(context, options.visionDelegate().get(), defaultBackend)
            : defaultBackend;
    Backend audioBackend =
        options.audioDelegate().isPresent()
            ? getBackendForDelegate(context, options.audioDelegate().get(), defaultBackend)
            : defaultBackend;

    Integer maxInputLength = options.maxInputLength().orElse(null);
    Integer visionTokensPerImage = options.visionTokensPerImage().orElse(null);

    String cacheDir = options.cacheDir().orElseGet(() -> context.getCacheDir().getAbsolutePath());

    EmbeddingEngineConfig config =
        new EmbeddingEngineConfig(
            finalPath,
            backend,
            visionBackend,
            audioBackend,
            cacheDir,
            fd,
            maxInputLength,
            visionTokensPerImage,
            options
                .activationDataType()
                .map(UniversalEmbedderOptions.ActivationDataType::toLiteRtLmType)
                .orElse(null));
    this.engine = new EmbeddingEngine(config);
    this.engine.initialize();
  }

  private static Backend getBackendForDelegate(
      Context context, Delegate delegate, Backend mainBackend) {
    switch (delegate) {
      case GPU:
        return new Backend.GPU();
      case NPU:
        String dispatchDir =
            mainBackend instanceof Backend.NPU
                ? ((Backend.NPU) mainBackend).getNativeLibraryDir()
                : (context.getApplicationInfo() != null
                        && context.getApplicationInfo().nativeLibraryDir != null
                    ? context.getApplicationInfo().nativeLibraryDir
                    : "");
        return new Backend.NPU(dispatchDir);
      case CPU:
      default:
        return new Backend.CPU();
    }
  }

  /**
   * Evaluates the selected acceleration configurations inside {@link BaseOptions} and maps them to
   * a native LiteRT-LM {@link Backend} with robust validation.
   */
  private static Backend getBackend(Context context, BaseOptionsProto.BaseOptions proto) {
    if (proto == null) {
      throw new MediaPipeException(
          MediaPipeException.StatusCode.INVALID_ARGUMENT.ordinal(),
          "BaseOptions proto cannot be null.");
    }

    try {
      if (proto.getAcceleration().hasGpu()) {
        return new Backend.GPU();
      }

      if (proto.getAcceleration().hasLitert()) {
        LiteRt litert = proto.getAcceleration().getLitert();

        if (litert.hasNpu()) {
          String dispatchDir =
              context.getApplicationInfo() != null
                      && context.getApplicationInfo().nativeLibraryDir != null
                  ? context.getApplicationInfo().nativeLibraryDir
                  : "";
          String customPath = litert.getNpu().getDispatchLibraryPath();
          if (!customPath.isEmpty()) {
            File customDir = new File(customPath);
            if (!customDir.exists()) {
              throw new MediaPipeException(
                  MediaPipeException.StatusCode.INVALID_ARGUMENT.ordinal(),
                  "NPU dispatch library directory does not exist: " + customPath);
            }
            dispatchDir = customPath;
          }
          return new Backend.NPU(dispatchDir);
        }

        if (litert.hasGpu()) {
          return new Backend.GPU();
        }

        if (litert.hasCpu()) {
          return new Backend.CPU();
        }

        return new Backend.CPU();
      }

      return new Backend.CPU();
    } catch (MediaPipeException e) {
      throw e;
    } catch (RuntimeException e) {
      MediaPipeException exception =
          new MediaPipeException(
              MediaPipeException.StatusCode.INTERNAL.ordinal(),
              "Failed to configure backend execution engine: " + e.getMessage());
      exception.initCause(e);
      throw exception;
    }
  }

  public static UniversalEmbedder createFromOptions(
      Context context, UniversalEmbedderOptions options) {
    return new UniversalEmbedder(context, options);
  }

  public EmbeddingResult embedText(String text) {
    ImmutableList<InputData> contents = ImmutableList.of(new InputData.Text(text));
    return executeInference(contents);
  }

  public EmbeddingResult embedImage(MPImage image) {
    Bitmap bitmap = BitmapExtractor.extract(image);
    ImmutableList<InputData> contents = ImmutableList.of(new InputData.Image(encodeAsTga(bitmap)));
    return executeInference(contents);
  }

  public EmbeddingResult embedAudio(AudioData audioData) {
    float[] buffer = audioData.getBuffer();
    AudioData.AudioDataFormat format = audioData.getFormat();
    byte[] wavBytes = encodeAsWav(buffer, (int) format.getSampleRate(), format.getNumOfChannels());
    ImmutableList<InputData> contents = ImmutableList.of(new InputData.Audio(wavBytes));
    return executeInference(contents);
  }

  /**
   * Performs embedding extraction on multi-modal parts of a content block.
   *
   * @param content the list of multi-modal content to embed.
   * @return a {@link EmbeddingResult} object that contains the embeddings.
   */
  public EmbeddingResult embedContent(List<Object> content) {
    return executeInference(convertToInputData(content));
  }

  public EmbeddingProvider getProvider() {
    return new EmbeddingProvider() {
      @Override
      public float[] embedContent(List<Object> content) {
        EmbeddingResult result = UniversalEmbedder.this.embedContent(content);
        if (result.embeddings().isEmpty()) {
          return new float[0];
        }
        return result.embeddings().get(0).floatEmbedding();
      }
    };
  }

  private List<InputData> convertToInputData(List<Object> content) {
    List<InputData> contents = new ArrayList<>();
    for (Object obj : content) {
      if (obj instanceof String) {
        contents.add(new InputData.Text((String) obj));
      } else if (obj instanceof byte[]) {
        contents.add(new InputData.Image((byte[]) obj));
      } else if (obj instanceof MPImage) {
        Bitmap bitmap = BitmapExtractor.extract((MPImage) obj);
        contents.add(new InputData.Image(encodeAsTga(bitmap)));
      } else if (obj instanceof AudioData) {
        AudioData audioData = (AudioData) obj;
        float[] buffer = audioData.getBuffer();
        AudioData.AudioDataFormat format = audioData.getFormat();
        byte[] wavBytes =
            encodeAsWav(buffer, (int) format.getSampleRate(), format.getNumOfChannels());
        contents.add(new InputData.Audio(wavBytes));
      } else {
        contents.add(new InputData.Text(String.valueOf(obj)));
      }
    }
    return contents;
  }

  private EmbeddingResult executeInference(List<InputData> contents) {
    long timestamp = syntheticTimestamp.getAndIncrement();
    statsLogger.recordCpuInputArrival(timestamp);
    EmbeddingOptions embeddingOptions = new EmbeddingOptions(options.l2Normalize());
    EmbeddingResponse response = engine.computeEmbedding(contents, embeddingOptions);
    Embedding embedding =
        Embedding.create(response.getEmbedding(), new byte[0], 0, Optional.empty());
    statsLogger.recordInvocationEnd(timestamp);
    return EmbeddingResult.create(Collections.singletonList(embedding), Optional.empty());
  }

  /** Utility function to compute cosine similarity between two {@link Embedding} objects. */
  public static double cosineSimilarity(Embedding u, Embedding v) {
    return CosineSimilarity.compute(u, v);
  }

  /** Utility function to compute cosine similarity between two float arrays. */
  public static double cosineSimilarity(float[] vectorA, float[] vectorB) {
    Embedding embeddingA = Embedding.create(vectorA, new byte[0], 0, Optional.empty());
    Embedding embeddingB = Embedding.create(vectorB, new byte[0], 0, Optional.empty());
    return CosineSimilarity.compute(embeddingA, embeddingB);
  }

  private static byte[] encodeAsWav(float[] pcmData, int sampleRate, int channels) {
    int dataSize = pcmData.length * 2;
    ByteBuffer buffer = ByteBuffer.allocate(44 + dataSize).order(ByteOrder.LITTLE_ENDIAN);

    // 1. RIFF Header
    buffer.put("RIFF".getBytes());
    buffer.putInt(dataSize + 36);
    buffer.put("WAVEfmt ".getBytes(UTF_8));

    // 2. Format Chunk
    buffer.putInt(16); // Subchunk1Size (16 for PCM)
    buffer.putShort((short) 1); // AudioFormat (1 for PCM)
    buffer.putShort((short) channels); // NumChannels
    buffer.putInt(sampleRate); // SampleRate
    buffer.putInt(sampleRate * channels * 2); // ByteRate
    buffer.putShort((short) (channels * 2)); // BlockAlign
    buffer.putShort((short) 16); // BitsPerSample

    // 3. Data Chunk
    buffer.put("data".getBytes());
    buffer.putInt(dataSize);

    // 4. Convert and write PCM Audio Data
    for (float sample : pcmData) {
      short val = (short) Math.max(-32768, Math.min(32767, Math.round(sample * 32767.0f)));
      buffer.putShort(val);
    }

    return buffer.array();
  }

  private static byte[] encodeAsTga(Bitmap bitmap) {
    int width = bitmap.getWidth();
    int height = bitmap.getHeight();
    int[] pixels = new int[width * height];
    bitmap.getPixels(pixels, 0, width, 0, 0, width, height);

    byte[] tgaBytes = new byte[20 + pixels.length * 4];
    ByteBuffer buffer = ByteBuffer.wrap(tgaBytes).order(ByteOrder.LITTLE_ENDIAN);

    // TGA Header (18 bytes) + 2 dummy bytes for alignment
    tgaBytes[0] = 2; // ID Length = 2 bytes (acts as padding)
    tgaBytes[2] = 2; // Uncompressed True-color
    tgaBytes[12] = (byte) (width & 0xFF);
    tgaBytes[13] = (byte) ((width >> 8) & 0xFF);
    tgaBytes[14] = (byte) (height & 0xFF);
    tgaBytes[15] = (byte) ((height >> 8) & 0xFF);
    tgaBytes[16] = 32; // 32 bits per pixel
    tgaBytes[17] = 0x28; // Top-down, 8 alpha bits

    // Clear bytes 18, 19 (Dummy ID bytes for alignment)
    tgaBytes[18] = 0;
    tgaBytes[19] = 0;

    // Fast Write Pixels
    buffer.position(20);
    buffer.asIntBuffer().put(pixels);

    return tgaBytes;
  }

  @Override
  public void close() {
    statsLogger.logSessionEnd();
    engine.close();
  }
}
